File: src/main/java/me/wolfyscript/customcrafting/recipes/RecipeTypeIdResolver.java
/*
 *       ____ _  _ ____ ___ ____ _  _ ____ ____ ____ ____ ___ _ _  _ ____
 *       |    |  | [__   |  |  | |\/| |    |__/ |__| |___  |  | |\ | | __
 *       |___ |__| ___]  |  |__| |  | |___ |  \ |  | |     |  | | \| |__]
 *
 *       CustomCrafting Recipe creation and management tool for Minecraft
 *                      Copyright (C) 2021  WolfyScript
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.wolfyscript.customcrafting.recipes;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.DatabindContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;
import com.fasterxml.jackson.databind.type.TypeFactory;
import me.wolfyscript.customcrafting.CustomCrafting;

public class RecipeTypeIdResolver extends TypeIdResolverBase {

    private JavaType superType;

    @Override
    public void init(JavaType baseType) {
        super.init(baseType);
        this.superType = baseType;
    }

    @Override
    public String idFromValue(Object value) {
        return getKey(value);
    }

    @Override
    public String idFromValueAndType(Object value, Class<?> aClass) {
        return getKey(value);
    }

    private String getKey(Object value) {
        if (value instanceof CustomRecipe<?> customRecipe) {
            return customRecipe.type.toString();
        } else {
            throw new IllegalArgumentException(String.format("Object %s is not of type CustomRecipe!", value.getClass().getName()));
        }
    }

    @Override
    public JavaType typeFromId(DatabindContext context, String id) {
        RecipeType<?> recipeType = CustomCrafting.inst().getRegistries().getRecipeTypes().get(CustomCrafting.inst().getApi().getIdentifiers().getNamespaced(id));
        if (recipeType != null) {
            Class<?> clazz = recipeType.getRecipeClass();
            return clazz != null ? context.constructSpecializedType(this.superType, clazz) : TypeFactory.unknownType();
        }
        return TypeFactory.unknownType();
    }

    @Override
    public JsonTypeInfo.Id getMechanism() {
        return JsonTypeInfo.Id.CUSTOM;
    }
}

File: api/src/main/kotlin/com/wolfyscript/customcrafting/recipes/RecipeTypeIdResolver.kt
package com.wolfyscript.customcrafting.recipes

import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.DatabindContext
import com.fasterxml.jackson.databind.DeserializationConfig
import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.SerializationConfig
import com.fasterxml.jackson.databind.jsontype.NamedType
import com.fasterxml.jackson.databind.jsontype.TypeDeserializer
import com.fasterxml.jackson.databind.jsontype.TypeSerializer
import com.fasterxml.jackson.databind.jsontype.impl.StdTypeResolverBuilder
import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase
import com.fasterxml.jackson.databind.type.TypeFactory
import com.wolfyscript.customcrafting.CustomCraftingProvider
import com.wolfyscript.customcrafting.util.CUSTOMCRAFTING_NAMESPACE
import com.wolfyscript.scafall.identifier.Key

class RecipeTypeIdResolver : TypeIdResolverBase() {

    private lateinit var superType: JavaType

    override fun init(bt: JavaType?) {
        super.init(bt)
        superType = bt ?: throw IllegalArgumentException("Failed to initialize recipe type resolver: missing super type!")
    }

    override fun idFromValue(value: Any?): String? {
        return getKey(value)
    }

    override fun idFromValueAndType(value: Any?, suggestedType: Class<*>?): String? {
        return getKey(value)
    }

    private fun getKey(value: Any?): String {
        if (value == null) {
            throw IllegalArgumentException("Failed to get recipe type null type!")
        }
        if (value is CustomRecipe<*,*>) {
            val key = CustomCraftingProvider.get().registries.recipeTypes.getKey(value.type)
            if (key != null) {
                return key.toString()
            }
            throw IllegalArgumentException("Failed to get recipe type key of recipe class: ${value::class.java}! Make sure the type it uses is registered.")
        }
        throw IllegalArgumentException("Failed to get recipe type of class: ${value::class.java}! That is not a recipe.")
    }

    override fun typeFromId(context: DatabindContext, id: String): JavaType {
        val namespacedKey = if (id.contains(':')) {
            Key.parse(id)
        } else {
            Key.key(Key.CUSTOMCRAFTING_NAMESPACE, id)
        }
        val value = CustomCraftingProvider.get().registries.recipeTypes[namespacedKey]
        if (value != null) {
            return context.constructSpecializedType(superType, value.recipeClass)
        }
        return TypeFactory.unknownType()
    }

    override fun getMechanism(): JsonTypeInfo.Id? {
        return JsonTypeInfo.Id.CUSTOM
    }

}

class RecipeTypeResolver : StdTypeResolverBuilder() {

    override fun buildTypeSerializer(
        config: SerializationConfig,
        baseType: JavaType,
        subtypes: Collection<NamedType>
    ): TypeSerializer? {
        return if (useForType(baseType)) super.buildTypeSerializer(config, baseType, subtypes) else null
    }

    override fun buildTypeDeserializer(
        config: DeserializationConfig,
        baseType: JavaType,
        subtypes: Collection<NamedType>
    ): TypeDeserializer? {
        return if (useForType(baseType)) super.buildTypeDeserializer(config, baseType, subtypes) else null
    }

    fun useForType(t: JavaType): Boolean {
        return t.isTypeOrSubTypeOf(CustomRecipe::class.java)
    }

}